### DISCUSSION TITLE
My program can't copy files

### DISCUSSION THREAD

**Discussion by khalidnecron** [tauri-apps/tauri#7320] [Q&A] [answered]
I have set to true the copyFile permission in tauri.conf.json but still can't copy files.

The tauri conf: 
```json
{
  "build": {
    "beforeDevCommand": "pnpm dev",
    "beforeBuildCommand": "pnpm build",
    "devPath": "http://localhost:1420",
    "distDir": "../dist",
    "withGlobalTauri": false
  },
  "package": {
    "productName": "Cobblemon Pokedex Refactor",
    "version": "0.0.2"
  },
  "tauri": {
    "allowlist": {
      "all": false,
      "shell": {
        "all": false,
        "open": true
      },
      "dialog": {
        "all": false,
        "ask": false,
        "confirm": false,
        "message": false,
        "open": true,
        "save": false
      },
      "fs": {
        "all": false,
        "copyFile": true,
        "createDir": true,
        "exists": true,
        "readDir": true,
        "readFile": true,
        "removeDir": false,
        "removeFile": false,
        "renameFile": true,
        "scope": [],
        "writeFile": true
      }
    },
    "bundle": {
      "active": true,
      "targets": "all",
      "identifier": "com.khalidnecron.dev",
      "icon": [
        "icons/32x32.png",
        "icons/128x128.png",
        "icons/128x128@2x.png",
        "icons/icon.icns",
        "icons/icon.ico"
      ]
    },
    "security": {
      "csp": null
    },
    "windows": [
      {
        "fullscreen": false,
        "resizable": true,
        "title": "Cobblemon Pokedex Refactor",
        "width": 400,
        "height": 400
      }
    ]
  }
}
```

the rust code:
```rust
async fn create_dir_for_files(path: &String) -> () {
    let entries = fs::read_dir(path).expect("Failed to read the directory");

    let rustemon_client = rustemon::client::RustemonClientBuilder::default()
        .with_mode(rustemon::client::CacheMode::NoStore)
        .try_build()
        .unwrap();

    for entry in entries {
        if let Ok(entry) = entry {
            if let Ok(entry_file_type) = entry.file_type() {
                if entry_file_type.is_file() {
                    let name = entry.file_name().into_string().unwrap();

                    let name: Vec<&str> = name.split('_').collect();

                    let pokemon =
                        match rustemon::pokemon::pokemon::get_by_name(&name[1], &rustemon_client)
                            .await
                        {
                            Ok(p) => p,
                            Err(_) => continue,
                        };

                    let id = if pokemon.id < 1000 {
                        format!("0{}", pokemon.id)
                    } else {
                        pokemon.id.to_string()
                    };

                    let new_dir = format!("{}/{}_{}", path, &id, &name[1]);

                    fs::create_dir_all(&new_dir).expect("failed to create dir");

                    fs::copy(&entry.path(), &new_dir).expect("failed to copy contents");
                }
            }
        }
    }
}
```

The type of error is AccessDenied. I wanna know why this happens and how to fix it

**Comment by FabianLars** [comment]
> I have set to true the copyFile permission in tauri.conf.json

The allowlist is only about the js side.

Also, AccessDenied is an error that comes directly from the OS not tauri or rust. That said, can you post the full error and tell us which line it complains about? And i guess the path you're reading would be helpful too.

Lastly, a full reproduction example, that we can test on our machines too would probably be nice if we can't figure it out with the prior questions.

**Reply by khalidnecron** [reply]
The error is this:
```
thread 'tokio-runtime-worker' panicked at 'failed to copy contents: Os { code: 5, kind: PermissionDenied, message: "Acceso denegado." }', src\refactor.rs:80:55
```
the line that complains about is fs::copy

For the path is the place where is storage any addon folder that the Minecraft mod Cobblemon can accept + "/assets/cobblemon/bedrock/species" or "/assets/cobblemon/bedrock/resolvers/"

For reproduction the best i can think of is doing it with my program:

1. Download a Cobblemon addon ( for example: https://modrinth.com/resourcepack/moremons/version/0.4.1)
2. Extract it into a folder 
3. Clone the repo https://github.com/khalidnecron/cobblemon-pokedex-refactor
4. Run pnpm install in the repo
5. Run pnpm tauri dev
6. Use the browse button and select the folder that have the extracted content of the zip
7. Use the refactor button

And that will reproduce the error

**Reply by khalidnecron** [reply]
Another note is that the addon that i put for example will have another error but is because one folder name is not in the pokeapi form

you can ignore that error

**Reply by FabianLars** [reply]
Alrighty, thanks for that! Gonna try it on my machine tomorrow. Just to be sure: Which OS are you testing this on?

**Reply by khalidnecron** [reply]
Windows

**Reply by FabianLars** [reply, marked as answer]
The destination path for `copy` needs to include the filename. In your code it's only the dir name. It then tries to overwrite a dir with a file which in turn gives you the permission denied error.
When i add the file name to it i don't get any errors anymore.

**Reply by khalidnecron** [reply]
Could you give me the code of that?

And thanks for your help

**Reply by FabianLars** [reply]
i used a hardcoded `whatever.json` but i assume you want to keep the file name as-is in which case the imo simplest solution looks like this (tho you may wanna start looking into working with Path/PathBuf):
```rs
fs::copy(
    &entry.path(),
    &format!("{new_dir}/{}", &entry.file_name().to_string_lossy()),
)
.expect("failed to copy contents");
```

**Reply by khalidnecron** [reply]
Thanks for your help
